File: python/day11.py
# DAY11

OFFS = [(1, 0), (1, 1), (0, 1), (-1, 1), (-1, 0), (-1, -1), (0, -1), (1, -1)]

def flash(board, x, y):
    res = 1
    board[y][x] == 11
    for dx, dy in OFFS:
        cx, cy = x + dx, y + dy
        if cx >= 0 and cx < len(board[0]) and cy >= 0 and cy < len(board):
            board[cy][cx] += 1
            if board[cy][cx] == 10:
                res += flash(board, cx, cy)
    return res

def step(board):
    num_flashes = 0
    for j in range(len(board)):
        for i in range(len(board[0])):
            board[j][i] += 1
            if board[j][i] == 10:
                num_flashes += flash(board, i, j)
    for j in range(len(board)):
        for i in range(len(board[0])):
            if board[j][i] >= 10:
                board[j][i] = 0
    return num_flashes

def solution():
    board = [list(map(int, list(s))) for s in open("../data/11.txt").read().split()]

    res1 = 0
    board1 = [row[:] for row in board]
    for i in range(100):
        res1 += step(board1)

    res2 = 0
    while True:
        num_flashes = step(board)
        res2 += 1
        if num_flashes == len(board) * len(board[0]):
            break

    print(f"Answer 1: {res1}\nAnswer 2: {res2}")

File: rust/src/day11.rs
use crate::common;

type Board = Vec<Vec<u8>>;

const OFFS: [(i32, i32); 8] = [
    (1, 0),
    (1, 1),
    (0, 1),
    (-1, 1),
    (-1, 0),
    (-1, -1),
    (0, -1),
    (1, -1),
];

fn flash(board: &mut Board, x: i32, y: i32) -> usize {
    let (w, h) = (board[0].len() as i32, board.len() as i32);
    let mut res = 1;
    board[y as usize][x as usize] = 11;
    for (dx, dy) in OFFS {
        let (cx, cy) = (x + dx, y + dy);
        if cx >= 0 && cx < w && cy >= 0 && cy < h {
            board[cy as usize][cx as usize] += 1;
            if board[cy as usize][cx as usize] == 10 {
                res += flash(board, cx, cy);
            }
        }
    }
    res
}

fn step(board: &mut Board) -> usize {
    let mut num_flashes = 0;
    let (w, h) = (board[0].len(), board.len());
    for j in 0..h {
        for i in 0..w {
            board[j][i] += 1;
            if board[j][i] == 10 {
                num_flashes += flash(board, i as i32, j as i32);
            }
        }
    }
    for j in 0..h {
        for i in 0..w {
            if board[j][i] >= 10 {
                board[j][i] = 0;
            }
        }
    }
    num_flashes
}

pub(crate) fn solution() {
    let lines = common::read_lines(&common::data_file(11)).unwrap();
    let board: Board = lines
        .iter()
        .map(|x| x.as_bytes().iter().map(|c| c - '0' as u8).collect())
        .collect();
    let (w, h) = (board[0].len(), board.len());

    let mut res1 = 0;
    let mut board1 = board.clone();
    for _ in 0..100 {
        res1 += step(&mut board1);
    }

    let mut res2 = 0;
    let mut board2 = board.clone();
    loop {
        let num_flashes = step(&mut board2);
        res2 += 1;
        if num_flashes == w * h {
            break;
        }
    }
    println!("Answer 1: {:?}\nAnswer 2: {:?}", res1, res2);
}
